File: 20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if not len(s):
            return True

        if len(s) == 1:
            return False

        stack = []
        pairs = {
            "(": ")",
            "{": "}",
            "[": "]",
        }

        for i in s:
            if i == "[" or i == "{" or i == "(":
                stack.append(i)
                continue
            if 0 == len(stack) or i != pairs[stack.pop()]:
                return False

        return len(stack) == 0


test_cases = [
    ("()", True),
    ("()[]{}", True),
    ("(]", False),
    ("([)]", False),
    ("{[]}", True),
    ("", True),  # empty string is valid
    ("(", False),  # single opening
    (")", False),  # single closing
    ("(()", False),  # missing closing
    ("())", False),  # extra closing
    ("((()))", True),  # nested same type
    ("{[()]}", True),  # nested mixed types
    ("{[(])}", False),  # incorrect nesting
    ("[]{}()", True),  # sequential valid pairs
    ("[", False),
    ("]", False),
    ("(((((((((())))))))))", True),  # deep nesting
    ("(((((((((()))))))))))", False),  # one extra closing
    ("()(()", False),
    ("(()())", True),
    ("{", False),
    ("}", False),
]

solution = Solution()
passed = 0

for i, (input_s, expected) in enumerate(test_cases, 1):
    result = solution.isValid(input_s)
    ok = result == expected

    print(
        f"Test {i:2d}: in: {repr(input_s)} → exp: {expected} | res: {result} :: "
        f"{'PASSED' if ok else 'FAILED'}"
    )

    if ok:
        passed += 1
    else:
        print("First failure — stopping.")
        break

print(f"\nSummary: {passed} / {len(test_cases)} tests passed")

File: 20-valid-parentheses.c
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

bool isValid(const char *s)
{
    char stack[10001];
    int top = 0;
    for (; *s; s++)
    {
        char c = *s;
        if (c == '(' || c == '[' || c == '{')
        {
            stack[top++] = c;
        }
        else
        {
            if (top == 0)
                return false;
            char open = stack[--top];
            // Valid pairs XOR to 1 or 6:
            // 'char' XOR 'char' = ASCII_CODE XOR ASCII_CODE
            // ')' ^ '(' = 41 ^ 40 = 1
            // ']' ^ '[' = 93 ^ 91 = 6
            // '}' ^ '{' = 125 ^ 123 = 6
            if ((c ^ open) != 6 && (c ^ open) != 1)
                return false;
        }
    }
    return top == 0;
}

bool isValid_original(const char *s)
{
    if (!s[0])
        return true;

    int n = 0;
    while (s[n])
        n++;
    if (n & 1)
        return false;

    char stack[n];
    int top = 0;

    for (int i = 0; i < n; i++)
    {
        char c = s[i];
        if (c == '(')
        {
            stack[top++] = ')';
        }
        else if (c == '[')
        {
            stack[top++] = ']';
        }
        else if (c == '{')
        {
            stack[top++] = '}';
        }
        else
        {
            if (top == 0 || stack[--top] != c)
            {
                return false;
            }
        }
    }

    return top == 0;
}

#ifndef MAIN_BENCH
int main(void)
{
    struct
    {
        const char *input;
        bool expected;
    } tests[] = {
        {"()", true},
        {"()[]{}", true},
        {"(]", false},
        {"([)]", false},
        {"{[]}", true},
        {"", true},
        {"(", false},
        {")", false},
        {"(()", false},
        {"())", false},
        {"((()))", true},
        {"{[()]}", true},
        {"{[(])}", false},
        {"[]{}()", true},
        {"[", false},
        {"]", false},
        {"(((((((((())))))))))", true},
        {"(((((((((()))))))))))", false},
        {"()(()", false},
        {"(()())", true},
        {"{", false},
        {"}", false},
        {"){", false},
    };

    const int n_tests = sizeof(tests) / sizeof(tests[0]);
    int passed = 0;

    for (int i = 0; i < n_tests; i++)
    {
        bool result = isValid(tests[i].input);

        bool ok = (result == tests[i].expected);

        printf("Test %2d: %-22s → exp: %-5s | res: %-5s  → %s\n", i + 1, tests[i].input[0] ? tests[i].input : "(empty)",
               tests[i].expected ? "true" : "false", result ? "true" : "false", ok ? "PASSED" : "FAILED");

        if (ok)
        {
            passed++;
        }
        else
        {
            printf("   ↑ first failure — stopping early\n");
            break;
        }
    }

    printf("\nSummary: %d / %d tests passed\n", passed, n_tests);

    return 0;
}
#endif
